File: python/divide_conquer/merge_k_sorted_lists.py
from typing import List, Optional

import heapq

class ListNode:
    def __init__(self, val:int = 0, next: 'ListNode' = None) -> None:
        self.val = val
        self.next = next

    def __repr__(self):
        return f"{self.val} -> {self.next}" if self.next else f"{self.val}"



class MinHeap:
    def __init__(self):
        self.heap = []

    def push(self, node: ListNode):
        heapq.heappush(self.heap, (node.val, node))

    def pop(self) -> ListNode:
        return heapq.heappop(self.heap)[1]

    def __len__(self) -> int:
        return len(self.heap)



class Solution:
    def merge_k_lists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        min_heap = MinHeap()


        for list_head in lists:
            if list_head:
                min_heap.push(list_head)


        dummy = ListNode()
        current = dummy

        while len(min_heap) > 0:
            node = min_heap.pop()

            current.next = node
            current = current.__new__

            if node.next:
                min_heap.push(node.next)


        return dummy.next


    def print_list(self, heap: ListNode):
        while heap:
            print(heap.val, end=" -> " if heap.next else "\n" )

            heap = heap.next

    def create_list(self, arr: List[int]) -> ListNode:
        dummy = ListNode()

        current = dummy

        for val in arr:
            current.next = ListNode(val)

            current = current.next

        return dummy.next

File: golang/divide_conquer/merge_k_sorted_lists.go
package main

import (
	"container/heap"
	"fmt"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

type MinHeap []*ListNode

func (h MinHeap) Len() int           { return len(h) }
func (h MinHeap) Less(i, j int) bool { return h[i].Val < h[j].Val }
func (h MinHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h *MinHeap) Push(x interface{}) {
	*h = append(*h, x.(*ListNode))
}

func (h *MinHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

func mergeKLists(lists []*ListNode) *ListNode {
	h := &MinHeap{}

	heap.Init(h)

	for _, list := range lists {
		if list != nil {
			heap.Push(h, list)
		}
	}

	dummy := &ListNode{}

	current := dummy

	for h.Len() > 0 {
		node := heap.Pop(h).(*ListNode)

		current.Next = node
		current = current.Next

		if node.Next != nil {
			heap.Push(h, node.Next)
		}
	}

	return dummy.Next
}

func printList(head *ListNode) {
	for head != nil {
		fmt.Print(head.Val, " ")
		head = head.Next
	}

	fmt.Println()
}

func createList(arr []int) *ListNode {
	dummy := &ListNode{}
	current := dummy

	for _, val := range arr {
		current.Next = &ListNode{Val: val}
		current = current.Next
	}

	return dummy.Next
}
